### PR TITLE
Build Mantid with gsl2

### LINKED ISSUE #15421: Mantid doesn't build with GSL 2
Mantid fails to build with GSL 2 which will be on Ubuntu 16.04 and Fedora 24. We will need to support Ubuntu 16.04 for the `3.7` release.

We have a Ubuntu 16.04 [build server](http://builds.mantidproject.org/job/master_clean-ubuntu-16.04/). Feel free to point it at your working branch to use for testing.

The type of errors that are appearing:
```
02:04:46 /home/builder/jenkins-linode/workspace/master_clean-ubuntu-16.04/Framework/CurveFitting/src/Algorithms/Fit1D.cpp: In member function ‘virtual void Mantid::CurveFitting::Algorithms::Fit1D::exec()’:
02:04:46 /home/builder/jenkins-linode/workspace/master_clean-ubuntu-16.04/Framework/CurveFitting/src/Algorithms/Fit1D.cpp:615:29: error: ‘struct gsl_multifit_fdfsolver’ has no member named ‘J’
02:04:46        gsl_multifit_covar(s->J, 0.0, covar);
```

### PR DESCRIPTION
**Pull request by mantid-roman**
Some changes are made to make it work with GSL2. Most of the unit tests are fixed. At least one test still fails (`OptimizeCrystalPlacement`).

**To test:**

Build on Ubuntu 16.04.

Fixes #15421.

It is too early for the release notes.

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by AndreiSavici** [comment]
Unit tests still failing:
	1163 - CurveFittingTest_MultiDomainFunctionTest (Failed)
	1200 - CrystalTest_OptimizeCrystalPlacementTest (Failed)
Doc test failed:
EnggCalibrate-v1.rst
```
Expected:
    DIFA1: 0
    DIFC1: 18267
    TZERO1: 277
    The output table has 1 row(s)
    Parameters from the table, DIFC1: 18267, ZERO1: 277
    Output GSAS iparam file was written? True
    Number of lines of the GSAS iparam file: 36
Got:
    DIFA1: 0
    DIFC1: 18266
    TZERO1: 278
    The output table has 1 row(s)
    Parameters from the table, DIFC1: 18266, ZERO1: 278
    Output GSAS iparam file was written? True
    Number of lines of the GSAS iparam file: 36
```

**Comment by AndreiSavici** [comment]
For the EnggCalibrate doctest, the TZERO1 is 277.51 plus or minus 3.0, and DIFC1 is 18267.05 plus or minus 1.72, so it's well within the error bars

**Comment by AndreiSavici** [comment]
:shipit: It should get into master. More fixes to follow

File: Framework/Crystal/test/IntegratePeakTimeSlicesTest.h [hunk 1]
```diff
@@ -190,8 +190,14 @@ class IntegratePeakTimeSlicesTest : public CxxTest::TestSuite {
           algP.getProperty("OutputWorkspace");
 
       TS_ASSERT_LESS_THAN(fabs(intensity - 60300), 1500.0);
-      // Not sure why this reduced the error so much in the test
-      TS_ASSERT_LESS_THAN(fabs(sigma - 457.0), 21.0);
+      // RT: my understanding is that there are 2 close minima
+      // that give different fitting errors which leads to changes in sigma.
+      if (sigma > 300.0) {
+        // Not sure why this reduced the error so much in the test
```
Comment (quantumsteve) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: Should this comment be added to lines 193-194 or removed?

Comment (mantid-roman) [inline Framework/Crystal/test/IntegratePeakTimeSlicesTest.h]: The `// RT:` comment tries to explain why the `if` is needed here.

**Comment by FedeMPouzols** [comment]
Good!

The EnggCalibrate will need coarser rounding it seems.

**Pull request merged**